Only curate external non-symbols in different lang

snprajwal opened the pull request:
Bug/issue #, if applicable: rdar://155522179

## Summary

PR https://github.com/swiftlang/swift-docc/pull/757 enabled DocC to include manually curated non-symbol nodes in the topics section, irrespective of the node language. This allows curating articles across documentation in different languages, even though the article's language is considered to be "Swift". However, the logic was too permissive, and allowed curating non-symbol nodes within the same bundle but across different languages. This is incorrect, since the target page should be annotated with the `SupportedLanguage` directive to enable curating it in a different language. A bespoke mechanism to curate an internal reference correctly is not necessary. This patch tightens the logic to only allow curating external references in a different language than the source page language, and removes the existing test for the undesired behaviour of including local references to cross-language symbols.

The changes in PR https://github.com/swiftlang/swift-docc/pull/757 were not propagated to the navigator, resulting in the references being present in the topics section but not in the sidebar. This patch also introduces the curation logic in the navigator to maintain parity with the topics section. The navigator tests for external render nodes have been modified accordingly to reflect this change.

## Dependencies

None

## Testing

The navigator tests in `ExternalRenderNodeTests` have been updated to reflect the curation behaviour.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

snprajwal commented:
@swift-ci please test

d-ronnqvist commented on Tests/SwiftDocCTests/Indexing/NavigatorIndexTests.swift:
```diff
@@ -894,6 +894,32 @@ Root
+                .first { $0.item.title == "Swift" }?
+                .children
+                .first { $0.item.title == "MixedLanguageFramework" }?
+                .children
+                .first { $0.item.title == "SwiftOnlyStruct2" }?
+                .children
+                .contains { $0.item.title == "ObjectiveCOnlyClass" },
```
This behavior doesn't seem correct to me. 

If I was browsing the Swift version of this documentation I would expect to not see any symbols that don't have Swift representations. 

Similarly, if I was browsing the Objective-C version of this documentation I would expect to not see any symbols that don't have an Objective-C representation. Curating a symbol under a page that it has no overlapping languages with, like this, feels like something we should raise a warning about if we don't already. 

IMO if a symbol is _only_ curated in a location that is has no overlapping languages with, we should treat it as uncrated and fall back to the automatic default curation for that symbol.

d-ronnqvist reviewed (changes requested):
Unless there's some information that I'm missing, this change in behavior feels like it's incorrect to me.

snprajwal committed:
Minor fixes to grammar and documentation

snprajwal commented:
@swift-ci please test

snprajwal commented:
@swift-ci please test

snprajwal committed:
Only curate external non-symbols in different lang

PR #757 enabled DocC to include manually curated non-symbol nodes in the
topics section, irrespective of the node language. This allows curating
articles across documentation in different languages, even though the
article's language is considered to be "Swift". However, the logic was
too permissive, and allowed curating non-symbol nodes within the same
bundle but across different languages. This is incorrect, since the
target page should be annotated with the `SupportedLanguage` directive
to enable curating it in a different language. A bespoke mechanism to
curate an internal reference correctly is not necessary. This patch
tightens the logic to only allow curating external references in a
different language than the source page language, and removes the
existing test for the undesired behaviour of including local references
to cross-language symbols.

The changes in PR #757 were not propagated to the navigator, resulting
in the references being present in the topics section but not in the
sidebar. This patch also introduces the curation logic in the navigator
to maintain parity with the topics section. The navigator tests for
external render nodes have been modified accordingly to reflect this
change.

rdar://155522179

snprajwal commented:
@swift-ci please test

d-ronnqvist reviewed (commented):
The changes look good. I think it would be good to add code comments in a few places to highlight that this behavior is a workaround, rather than the intended behavior.

d-ronnqvist reviewed (approved):
LGTM!
Thank you for working through all the feedback.

d-ronnqvist committed:
Merge branch 'main' into pnadig/fix-nav

d-ronnqvist commented:
@swift-ci please test
